lib/charms/mysql/v0/async_replication.py

```diff
@@ -481,53 +574,67 @@ def _on_primary_relation_changed(self, event):
             # Recover replica cluster
             self._charm.unit.status = MaintenanceStatus("Replica cluster in recovery")
 
+    def _on_offer_relation_broken(self, event: RelationBrokenEvent):
+        """Handle the async_primary relation being broken."""
+        if self._charm.unit.is_leader():
+            # remove relation secret by id
+            if secret_id := event.relation.data[self.model.app].get("secret-id"):
+                logger.debug("Removing replication secret")
+                secret = self._charm.model.get_secret(id=secret_id)
+                secret.remove_all_revisions()
+            else:
+                logger.debug("Secret-id not set, skipping removal")
+
+    def _on_secret_change(self, event: SecretChangedEvent):
+        """Propagates the secret being changed."""
```

Comment:
yes.
It should but that add extra complexity since we always deploy an app with it own secrets, so unifying old need to delete one, and recreating on relation broken.
giving the troubled history with secrets and CMR, I rather do some duplication that works for now.

Reply:
It can't. Juju design it thinking in a saas model, where a service is provider. In practice the relation offer side is the only one that can manipulate the secret granted to the relation. The consumer side can neither write nor grant a secret to the relation.